macrostep.el

```diff
@@ -708,17 +711,20 @@ Also removes the overlay from `macrostep-overlays'."
     ;; Remove overlay from the list and delete it
     (setq macrostep-overlays
           (delq overlay macrostep-overlays))
+    (let ((back-overlay (overlay-get overlay 'back-overlay)))
+      (when back-overlay (delete-overlay back-overlay)))
     (delete-overlay overlay)))
 
 (defun macrostep-collapse-overlays-in (start end)
   "Collapse all macrostepper overlays that are strictly between START and END.
 
 Will not collapse overlays that begin at START and end at END."
   (dolist (ol (overlays-in start end))
-    (if (and (> (overlay-start ol) start)
-	     (< (overlay-end ol) end)
-	     (overlay-get ol 'macrostep-original-text))
-	(macrostep-collapse-overlay ol t))))
+    (when (and (overlay-buffer ol)        ; collapsing may delete other overlays
```

Comment:
Can you explain the additional `(overlay-buffer ol)` condition? I guess the potential bug is that we call `(overlays-in start end)` at the top of the loop and loop over the resulting list, but overlays that come later in the list may already have been deleted by collapsing overlays that come earlier. Good catch.

Is this required for the addition of the highlight-only overlays or is it more of a general cleanup/bugfix? If it's the latter, it might be good to put it in a separate commit with its own explanatory commit message.